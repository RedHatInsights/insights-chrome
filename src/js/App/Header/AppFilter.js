import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { Button } from '@patternfly/react-core/dist/js/components/Button/Button';
import { CaretDownIcon } from '@patternfly/react-icons/dist/js/icons/caret-down-icon';
import { Dropdown } from '@patternfly/react-core/dist/js/components/Dropdown/Dropdown';
import { DropdownToggle } from '@patternfly/react-core/dist/js/components/Dropdown/DropdownToggle';
import { EmptyState, EmptyStateVariant } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyState';
import { EmptyStateIcon } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyStateIcon';
import { EmptyStateBody } from '@patternfly/react-core/dist/js/components/EmptyState/EmptyStateBody';
import FilterIcon from '@patternfly/react-icons/dist/js/icons/filter-icon';
import { Flex } from '@patternfly/react-core/dist/js/layouts/Flex/Flex';
import { SearchInput } from '@patternfly/react-core/dist/js/components/SearchInput/SearchInput';
import { Split } from '@patternfly/react-core/dist/js/layouts/Split/Split';
import { SplitItem } from '@patternfly/react-core/dist/js/layouts/Split/SplitItem';
import { TextContent } from '@patternfly/react-core/dist/js/components/Text/TextContent';
import { Text } from '@patternfly/react-core/dist/js/components/Text/Text';
import { Title } from '@patternfly/react-core/dist/js/components/Title/Title';
import { Spinner } from '@patternfly/react-core/dist/js/components/Spinner/Spinner';
import { Bullseye } from '@patternfly/react-core/dist/js/layouts/Bullseye/Bullseye';

import './AppFilter.scss';
import useGlobalNav from '../../utils/useGlobalNav';

const App = ({ id, title, routes, parent }) => (
  <div className="galleryItem">
    <Split>
      <SplitItem>
        <TextContent>
          <Text component="h4">{title}</Text>
          {routes.map((subApp) => {
            const redirectUrl = `${parent ? `${parent.id}/` : ''}${subApp.reload || `${id}/${subApp.id}`}`;

            return (
              <Text component="p" key={`${id}/${subApp.id}`}>
                <Text component="a" href={redirectUrl}>
                  {subApp.title}
                </Text>
              </Text>
            );
          })}
        </TextContent>
      </SplitItem>
    </Split>
  </div>
);

App.propTypes = {
  id: PropTypes.string,
  title: PropTypes.node,
  routes: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string.isRequired })),
  parent: PropTypes.shape({
    id: PropTypes.string,
  }),
};

const AppFilterDropdown = ({ isLoaded, setIsOpen, isOpen, filterValue, setFilterValue, filteredApps }) => (
  <Dropdown
    className="ins-c-page__app-filter-dropdown"
    onSelect={() => setIsOpen(true)}
    toggle={
      <DropdownToggle id="toggle-id" onToggle={() => setIsOpen(!isOpen)} toggleIndicator={CaretDownIcon}>
        Find an app or service
      </DropdownToggle>
    }
    isOpen={isOpen}
    ouiaId="App Filter"
  >
    <div className="content">
      {isLoaded ? (
        <React.Fragment>
          <Flex className="search">
            <SearchInput
              data-ouia-component-id="app-filter-search"
              placeholder="Find application or service"
              value={filterValue}
              onChange={(val) => setFilterValue(val)}
              onClear={() => setFilterValue('')}
            />
          </Flex>
          {filteredApps?.length > 0 ? (
            <div className="gallery">
              {filteredApps.map((app) => (
                <App key={app.id} {...app} />
              ))}
            </div>
          ) : (
            <EmptyState className="pf-u-mt-xl" variant={EmptyStateVariant.full}>
              <EmptyStateIcon className="pf-u-mb-xl" icon={FilterIcon} />
              <Title headingLevel="h4">No matching applications or services found.</Title>
              <EmptyStateBody className="pf-u-mb-xl">
                This filter criteria matches no applications or services. Try changing your input filter.
              </EmptyStateBody>
              <Button ouiaId="app-filter-clear-input" className="pf-u-mt-lg" variant="link" onClick={() => setFilterValue('')}>
                Clear all filters
              </Button>
            </EmptyState>
          )}
        </React.Fragment>
      ) : (
        <Bullseye className="pf-u-p-xl">
          <Spinner />
        </Bullseye>
      )}
    </div>
  </Dropdown>
);

AppFilterDropdown.propTypes = {
  isLoaded: PropTypes.bool,
  isOpen: PropTypes.bool,
  filterValue: PropTypes.string,
  setFilterValue: PropTypes.func,
  setIsOpen: PropTypes.func.isRequired,
  filteredApps: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      title: PropTypes.node.isRequired,
      routes: PropTypes.arrayOf(PropTypes.shape({ id: PropTypes.string.isRequired })),
    })
  ),
};

const AppFilter = () => {
  const { filteredApps, isLoaded, isOpen, setIsOpen, filterValue, setFilterValue } = useGlobalNav();

  return (
    <React.Fragment>
      <AppFilterDropdown
        isLoaded={!!isLoaded}
        setIsOpen={setIsOpen}
        isOpen={isOpen}
        filterValue={filterValue}
        setFilterValue={setFilterValue}
        filteredApps={filteredApps}
      />
      <div className={classnames({ 'pf-c-backdrop': isOpen })} />
    </React.Fragment>
  );
};

export default AppFilter;
