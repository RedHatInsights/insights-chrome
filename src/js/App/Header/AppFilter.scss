@media screen and (max-width:768px) {
  .pf-c-page__header {
    &-tools-group {
      margin-left: 0 !important;
    }
    &-tools {
      padding-right: 0;
    }
  }
}

.pf-c-page__header {
  .pf-c-backdrop {
    z-index: -10;
  }
}

.ins-c-page__app-filter-dropdown {
  .pf-c-dropdown__toggle {
    display: inline-block;
    margin-left: 72px;
    margin-right: var(--pf-global--spacer--sm);
    position: relative;
    white-space: nowrap;
    &::before {
      --pf-c-dropdown__toggle--before--BorderTopColor: var(--pf-global--disabled-color--100);
      --pf-c-dropdown__toggle--before--BorderRightColor: var(--pf-global--disabled-color--100);
      --pf-c-dropdown__toggle--before--BorderLeftColor: var(--pf-global--disabled-color--100);
    }
  }
  @media screen and (max-width:768px) { 
      .pf-c-dropdown__toggle {
        margin-left: 10px;
      }
  }
  .pf-c-dropdown__menu::before {
    z-index: -1;
  }


  .pf-c-dropdown__menu {
    left: 0;
    top: 78px;
    overflow-y:scroll;
    overflow-x:hidden;
    position: fixed; // allows menu to go full width
    .content {
      display: flex;
      flex-direction: column;
      align-self: stretch;
      .search {
        padding: var(--pf-global--spacer--xl) 0 var(--pf-global--spacer--xl) 0;
        justify-content: center;
        .pf-c-search-input {
          display: flex;
          width: 300px;
          &__icon {
              color: var(--pf-global--BackgroundColor--dark-100);
          }
          &__text-input{
              color: var(--pf-global--BackgroundColor--dark-100);
          }
        }
      }
      div.gallery {
        column-gap: var(--pf-global--spacer--lg);
        margin: 0 auto; // centers columns
      }
      div.galleryItem {
        display: block;
        padding-bottom: var(--pf-global--spacer--xl);
        width: 350px;
        overflow: hidden; // prevents padding from spilling into next column
        break-inside: avoid-column; // prevents padding from spilling into next column
        .pf-l-split {
          margin-top: auto;
          .pf-l-split__item {
            h4 {
              color: var(--pf-global--BackgroundColor--dark-100);
              font: "RedHatDisplay", "Overpass", overpass, helvetica, arial, sans-serif; //--pf-global--FontFamily--heading--sans-serif
              font-weight:  var(--pf-global--FontWeight--normal);
              margin-bottom: var(--pf-global--spacer--md);
              white-space: nowrap;
            }
            p {
              margin-bottom: var(--pf-global--spacer--sm);
            }
            a {
              white-space: nowrap;
              color: var(--pf-global--active-color--100);
            }
          }
        }
      }
    }
  }
  @media screen and (max-width:768px) { 
    .pf-c-dropdown__menu {
      bottom:0; // allows scrolling
      div.gallery {
        column-count: 1;
      }
    }
  }
  @media screen and (min-width: 769px) and (max-width: 1099px) {
    .pf-c-dropdown__menu {
      bottom: 0;  // allows scrolling
      div.gallery {
        column-count: 2;
        width: calc( 350px * 2 );
      }
    }
  }
  @media screen and (min-width: 1100px) and (max-width: 1549px) { 
    div.gallery {
      column-count: 3;
      width: calc( 350px * 3 );
    }
  }

  @media screen and (min-width: 1550px) { 
    div.gallery {
      column-count: 4;
      width: calc( 350px * 4 );
    }
  }

  .pf-c-empty-state {
    --pf-c-empty-state__body--Color: #6a6e73;
    h4 {
      color: var(--pf-global--BackgroundColor--dark-100);
      font-size: var(--pf-c-title--m-xl--FontSize);
    }
    button {
      color: var(--pf-global--active-color--100);
    }
  }
}
