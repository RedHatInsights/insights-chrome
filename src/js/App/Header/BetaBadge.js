import React from 'react';
import PropTypes from 'prop-types';
import { Badge } from '@patternfly/react-core';
import classNames from 'classnames';
import { useIntl } from 'react-intl';
import messages from '../../Messages';

const BetaBadge = ({ className, ...props }) => {
  const intl = useIntl();

  return (
    <Badge className={classNames('chr-c-toolbar__beta-badge', className)} {...props}>
      {intl.formatMessage(messages.beta)}
    </Badge>
  );
};

BetaBadge.propTypes = {
  className: PropTypes.string,
};

export default BetaBadge;
