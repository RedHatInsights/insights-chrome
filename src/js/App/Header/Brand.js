import React from 'react';
import { connect } from 'react-redux';
import { onToggle } from '../../redux/actions';
import { Button } from '@patternfly/react-core/dist/esm/components/Button';
import BarsIcon from '@patternfly/react-icons/dist/esm/icons/bars-icon';
import Logo from './Logo';

const Brand = ({ toggleNav, navHidden }) => (
    <div className="pf-c-page__header-brand">
        <div hidden={navHidden} className='pf-c-page__header-brand-toggle'>
            <Button
                variant="plain"
                aria-label="Toggle primary navigation"
                widget-type="InsightsNavToggle"
                onClick={() => toggleNav && toggleNav()}
            >
                <BarsIcon size="md"/>
            </Button>
        </div>
        <a className="pf-c-page__header-brand-link" href="./">
            <Logo />
        </a>
    </div>
);

function mapDispatchToProps(dispatch) {
    return {
        toggleNav: () => {
            dispatch(onToggle());
        }
    };
}

export default connect(({ chrome: { navHidden } }) => ({ navHidden }), mapDispatchToProps)(Brand);
