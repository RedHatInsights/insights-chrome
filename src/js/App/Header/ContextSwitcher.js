import React, { useEffect, useState } from 'react';
import PropTypes from 'prop-types';
import { Bullseye, ContextSelector, ContextSelectorItem, Spinner, TextContent, Text } from '@patternfly/react-core';
import CheckIcon from '@patternfly/react-icons/dist/js/icons/check-icon';
import { useDispatch, useSelector } from 'react-redux';
import classNames from 'classnames';
import axios from 'axios';

import { onToggleContextSwitcher } from '../../redux/actions';

import './ContextSwitcher.scss';
import { Fragment } from 'react';
import Cookies from 'js-cookie';
import { ACTIVE_ACCOUNT_SWITCH_NOTIFICATION, ACTIVE_REMOTE_REQUEST, CROSS_ACCESS_ACCOUNT_NUMBER, REQUESTS_COUNT, REQUESTS_DATA } from '../../consts';

const ContextSwitcher = ({ user, className }) => {
  const dispatch = useDispatch();
  const isOpen = useSelector(({ chrome }) => chrome?.contextSwitcherOpen);
  const [data, setData] = useState([]);
  const [searchValue, setSearchValue] = useState('');
  const [selectedAccountNumber, setSelectedAccountNumber] = useState(user.identity.account_number);
  const onSelect = () => {
    dispatch(onToggleContextSwitcher());
  };

  const handleItemClick = (target_account, request_id, end_date) => {
    if (target_account === selectedAccountNumber) {
      return;
    }
    localStorage.removeItem(ACTIVE_ACCOUNT_SWITCH_NOTIFICATION);
    localStorage.removeItem(REQUESTS_COUNT);
    localStorage.removeItem(REQUESTS_DATA);
    setSelectedAccountNumber(target_account);
    Cookies.set(CROSS_ACCESS_ACCOUNT_NUMBER, target_account);
    /**
     * We need to keep the request id somewhere to check if the request is still active after session start.
     * If it is not active, we have to remove the cookie.
     * This has to happen before ANY API call is made.
     */
    localStorage.setItem(
      ACTIVE_REMOTE_REQUEST,
      JSON.stringify({
        request_id,
        target_account,
        end_date,
      })
    );
    localStorage.setItem(ACTIVE_ACCOUNT_SWITCH_NOTIFICATION, 'true');
    window.location.reload();
  };

  const resetAccountRequest = () => {
    if (user?.identity?.account_number === selectedAccountNumber) {
      return;
    }
    setSelectedAccountNumber(user?.identity?.account_number);
    Cookies.remove(CROSS_ACCESS_ACCOUNT_NUMBER);
    localStorage.removeItem(ACTIVE_REMOTE_REQUEST);
    window.location.reload();
  };

  useEffect(() => {
    const initialAccount = localStorage.getItem(ACTIVE_REMOTE_REQUEST);
    if (initialAccount) {
      try {
        setSelectedAccountNumber(JSON.parse(initialAccount).target_account);
      } catch {
        console.log('Unable to parse initial account. Using default account');
      }
    }
    axios
      .get('/api/rbac/v1/cross-account-requests/', {
        params: {
          status: 'approved',
          order_by: '-created',
          query_by: 'user_id',
        },
      })
      .then(({ data: { data } }) =>
        setData(
          data
            .reduce((acc, curr) => (acc.find(({ target_account }) => target_account === curr.target_account) ? acc : [...acc, curr]), [])
            .filter(({ target_account }) => target_account !== user.identity.account_number)
        )
      );
  }, []);

  if (data.length === 0) {
    return null;
  }

  const filteredData = data && data.filter(({ target_account }) => `${target_account}`.includes(searchValue));

  return (
    <ContextSelector
      toggleText={`Acct: ${selectedAccountNumber}`}
      className={classNames('ins-c-page__context-switcher-dropdown', className)}
      onSearchInputChange={(val) => setSearchValue(val)}
      isOpen={isOpen}
      searchInputValue={searchValue}
      onToggle={onSelect}
      onSelect={onSelect}
      screenReaderLabel="Selected account:`"
      ouiaId="Account Switcher"
      searchInputPlaceholder="Search account"
    >
      {user && user?.identity?.account_number.includes(searchValue) ? (
        <ContextSelectorItem onClick={resetAccountRequest}>
          <TextContent className="personal-account">
            <Text className="account-label pf-u-mb-0">
              <span>{user?.identity?.account_number}</span>
              {user?.identity?.account_number === `${selectedAccountNumber}` && (
                <CheckIcon size="sm" color="var(--pf-global--primary-color--100)" className="pf-u-ml-auto" />
              )}
            </Text>
            <Text component="small">Personal account</Text>
          </TextContent>
        </ContextSelectorItem>
      ) : (
        <Fragment />
      )}
      {filteredData?.length === 0 ? <ContextSelectorItem>No results</ContextSelectorItem> : <Fragment />}
      {filteredData ? (
        filteredData.map(({ target_account, request_id, end_date }) => (
          <ContextSelectorItem onClick={() => handleItemClick(target_account, request_id, end_date)} key={request_id}>
            {target_account}
            {target_account === selectedAccountNumber && (
              <CheckIcon size="sm" color="var(--pf-global--primary-color--100)" className="pf-u-ml-auto" />
            )}
          </ContextSelectorItem>
        ))
      ) : (
        <ContextSelectorItem>
          <Bullseye>
            <Spinner size="md" />
          </Bullseye>
        </ContextSelectorItem>
      )}
    </ContextSelector>
  );
};

ContextSwitcher.propTypes = {
  user: PropTypes.object,
  className: PropTypes.string,
};

export default ContextSwitcher;
