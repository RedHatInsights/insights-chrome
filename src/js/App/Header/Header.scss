.chr-c-masthead {
  overflow: visible;
  .chr-c-dropdown-item__stack {
    white-space: nowrap;
    * { color: var(--pf-c-dropdown__menu-item--disabled--Color); }
  }
  .chr-c-dropdown-item__stack--header,
  .chr-c-dropdown-item__stack--subValue {
    font-size: var(--pf-global--FontSize--xs);
  }
  .chr-c-dropdown-item__stack--subValue { padding-top: var(--pf-global--spacer--xs); }

  .chr-c-brand {
    width: 193px;
  }
  .chr-c-badge-beta {
    background-color: var(--pf-global--palette--green-300);
    color: var(--pf-global--Color--dark-100);
  }
  .chr-c-dropdown-user-toggle {
    button::before {
      border-color: var(--pf-global--palette--black-900)
    }
    &.pf-m-expanded button::before, button:hover::before {
      border-bottom-color: var(--pf-global--primary-color--200)
    }    
  }
}

.chr-pen-test {
  border-top: 3px solid var(--pf-global--palette--purple-300);
}

.disabled-pointer {
  pointer-events: none;
}

.visible-pointer {
  pointer-events: auto;
}