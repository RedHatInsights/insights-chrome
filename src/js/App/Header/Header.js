import React, { Fragment } from 'react';
import Brand from './Brand';
import Tools from './Tools';
import UnAuthtedHeader from './UnAuthtedHeader';
import { useSelector } from 'react-redux';

const Header = () => {
  const user = useSelector(({ chrome: { user } }) => user);
  return user ? (
    <Fragment>
      <Brand />
      <Tools />
    </Fragment>
  ) : (
    <UnAuthtedHeader />
  );
};

export default Header;
