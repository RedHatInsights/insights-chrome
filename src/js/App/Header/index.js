import React, { Fragment } from 'react';
import Brand from './Brand';
import Tools from './Tools';
import Login from './Login';

export function unauthed() {
    return <Fragment>
        <Brand />
        <Login />
    </Fragment>;
}

;

const Header = () => (
    <Fragment>
        <Brand />
        <Tools />
    </Fragment>
);

export default Header;
