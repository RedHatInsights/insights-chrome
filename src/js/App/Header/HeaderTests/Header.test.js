import React from 'react';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import { Header } from '../Header';
import UnauthedHeader from '../UnAuthtedHeader';

describe('Header', () => {
  let initialState;
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore();
    initialState = {
      chrome: {
        activeTechnology: 'someTechnology',
        activeLocation: 'someLocation',
        user: {},
        navigation: {},
      },
    };
  });
  it('should render correctly', () => {
    const store = mockStore(initialState);
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <Header />
        </Provider>
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });
});

describe('unauthed', () => {
  let initialState;
  let mockStore;

  beforeEach(() => {
    mockStore = configureStore();
    initialState = {
      chrome: {
        activeTechnology: 'someTechnology',
        activeLocation: 'someLocation',
      },
    };
  });

  it('should render correctly', () => {
    const store = mockStore(initialState);
    const { container } = render(
      <MemoryRouter>
        <Provider store={store}>
          <UnauthedHeader />
        </Provider>
      </MemoryRouter>
    );
    expect(container.querySelector('div')).toMatchSnapshot();
  });
});
