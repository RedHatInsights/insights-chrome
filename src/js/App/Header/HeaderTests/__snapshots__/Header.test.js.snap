// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header should render correctly 1`] = `
<div>
  <a
    class="ins-m-hide-on-md ins-c-header-link"
    href="./"
  >
    <img
      alt="Red Hat Logo"
      src="test-file-stub"
    />
  </a>
  <div
    class="pf-c-dropdown ins-c-page__app-filter-dropdown-toggle"
    data-ouia-component-id="App Filter"
    data-ouia-component-type="PF4/Dropdown"
    data-ouia-safe="true"
  >
    <button
      aria-expanded="false"
      aria-haspopup="false"
      class="pf-c-dropdown__toggle"
      data-ouia-component-id="OUIA-Generated-DropdownToggle-1"
      data-ouia-component-type="PF4/DropdownToggle"
      data-ouia-safe="true"
      id="toggle-id"
      type="button"
    >
      <span
        class="pf-c-dropdown__toggle-text"
      >
        All apps and services
      </span>
      <span
        class="pf-c-dropdown__toggle-icon"
      >
        <svg
          aria-hidden="true"
          fill="currentColor"
          height="1em"
          role="img"
          style="vertical-align: -0.125em;"
          viewBox="0 0 320 512"
          width="1em"
        >
          <path
            d="M31.3 192h257.3c17.8 0 26.7 21.5 14.1 34.1L174.1 354.8c-7.8 7.8-20.5 7.8-28.3 0L17.2 226.1C4.6 213.5 13.5 192 31.3 192z"
          />
        </svg>
      </span>
    </button>
  </div>
  <div
    class=""
  />
</div>
`;

exports[`unauthed should render correctly 1`] = `
<div
  class="pf-c-page__header-tools"
>
  <button
    aria-disabled="false"
    aria-label="Toggle primary navigation"
    class="pf-c-button pf-m-tertiary"
    data-ouia-component-id="top-right-login-button"
    data-ouia-component-type="PF4/Button"
    data-ouia-safe="true"
    type="button"
    widget-type="InsightsNavToggle"
  >
    Log in
  </button>
</div>
`;
